### ISSUE TITLE
Continuous toggle ON/OFF when auto insert armed home

### ISSUE THREAD

**Issue by andyinno** [runningman84/home-assistant-apps#10]
Hi!
I updated to the new version of your app. At the moment everything is running smooth except that at the time I set for the auto arm home the system became crazy and start changing it's status in armed_home, disarmed.
I tried to change the behaviour of the app changing those 2 line of code.

    self.listen_state(self.alarm_disarm_auto_callback, sensor, old="not_home", new="home", duration=i)
    self.listen_state(self.alarm_arm_home_auto_state_change_callback, sensor, old="not_home", new="home", duration=15 * 60 + 1)
            i += 1

Basically I added the old sstatus, obviously the second one does not start anymore because it can happen that I am at home for more than 15 minutes.
Did you observe anything strange with those callbacks?
A better way to solve the continuous state toggle?

Thank you!

**Comment by runningman84** [comment]
Please provide the log output... I have not seen the same behavior before.

**Comment by andyinno** [comment]
Hi,
This is an example. I don't think it is related to #11 because I don't see the changes not_home => home but only triggers like home => home
This is a partial log by the way. What I see at home is that the alarm start toggling from armed_home to disarmed to armed_home.

    2019-10-29` 23:36:00.503384 INFO alarm: Callback alarm_arm_home_auto_timer
    2019-10-29 23:36:00.510236 INFO alarm: Running alarm_arm_home_auto
    2019-10-29 23:36:00.518412 INFO alarm: Calling service alarm_control_panel/alarm_arm_home
    2019-10-29 23:36:00.598473 INFO alarm: Callback alarm_state_armed_home from alarm_control_panel.home_alarm:state disarmed->armed_home2019-10-29 23:36:54.503540 INFO alarm: Callback alarm_disarm_auto from person.andrea:state home->home
    2019-10-29 23:36:54.509872 INFO alarm: Calling service alarm_control_panel/alarm_disarm
    2019-10-29 23:36:54.590080 INFO alarm: Callback alarm_state_disarmed from alarm_control_panel.home_alarm:state armed_home->disarmed
    2019-10-29 23:37:07.501953 INFO alarm: Callback alarm_arm_home_auto_device_change from person.micky:state home->home
    2019-10-29 23:37:07.508452 INFO alarm: Running alarm_arm_home_auto
    2019-10-29 23:37:07.516629 INFO alarm: Calling service alarm_control_panel/alarm_arm_home
    2019-10-29 23:37:07.581662 INFO alarm: Callback alarm_state_armed_home from alarm_control_panel.home_alarm:state disarmed->armed_home
    2019-10-29 23:37:21.498830 INFO alarm: Callback alarm_disarm_auto from person.andrea:state home->home
    2019-10-29 23:37:21.502088 INFO alarm: Calling service alarm_control_panel/alarm_disarm
    2019-10-29 23:37:21.567756 INFO alarm: Callback alarm_state_disarmed from alarm_control_panel.home_alarm:state armed_home->disarmed
    2019-10-29 23:37:24.502952 INFO alarm: Callback alarm_disarm_auto from person.micky:state home->home
    2019-10-29 23:37:24.511273 INFO alarm: Ignoring status home of person.micky because alarm system is in state disarmed
    2019-10-29 23:37:58.502446 INFO alarm: Callback alarm_arm_home_auto_device_change from person.andrea:state home->home
    2019-10-29 23:37:58.508634 INFO alarm: Running alarm_arm_home_auto
    2019-10-29 23:37:58.512838 INFO alarm: Calling service alarm_control_panel/alarm_arm_home
    2019-10-29 23:37:58.583530 INFO alarm: Callback alarm_state_armed_home from alarm_control_panel.home_alarm:state disarmed->armed_home
    2019-10-29 23:39:37.502266 INFO alarm: Callback alarm_arm_home_auto_device_change from person.micky:state home->home
    2019-10-29 23:39:37.508319 INFO alarm: Running alarm_arm_home_auto
    2019-10-29 23:39:37.511952 INFO alarm: Ignoring because alarm system is in state armed_home

**Comment by runningman84** [comment]
Home -> Home is not an event I am listening for.... Right now I don't know why you get this kind of error

**Comment by andyinno** [comment]
Well... I have an idea, but I am unsure about it.
I use https://github.com/andrewjfreyer/presence as bluetooth tracker, it gives me a confidence level about being at home or not.
Then I use the confidence level to transform it in home/not_home status like here: https://github.com/andrewjfreyer/presence#example-use-with-home-assistant

I am thinking that using this may be different from a real tracker, and it updates the status "home" also if I am at already at home. Like:
- confidence 100% => Home
- confidence 90%
- confidence 80%
- confidence 100% => Home

and this triggers the update.
I would like to change the listen_state from new=home to something like old=not_home new=home but I am unsure about which listen_state funcion I have to change.
As they are now, they can trigger also if the old_state is home.

**Comment by runningman84** [comment]
Ah ok please check line 111 and 112 and just add an old condition

**Comment by runningman84** [comment]
I have just added the condition to the alarm.py code please try and give me some feedback...

**Comment by andyinno** [comment]
Hi!
I tested it for a couple of days.

I can confirm that the changes is working.

In my opinion it is nicer with this configuration:

    self.listen_state(self.alarm_disarm_auto_callback, sensor, old="not_home", new="home", duration=i)
    self.listen_state(self.alarm_arm_home_auto_state_change_callback, sensor, new="home", duration=15 * 60 + 1)

What I see happening is:

At the time defined if I was already at home it toggles to home mode but it does not switch back to disabled.
Adding the not_home option to the second line I need to enter home after the time for trigger the function. I think that the expected behavior is to trigger the function also if you are already at home. Am I wrong?

**Comment by runningman84** [comment]
Using both old and new attributes just helps to ensure that only real state changes trigger something. Otherwise an updated home status (maybe due to change in confidence level) would also trigger something.

The second line is for this use case, nobody is home at alarm_arm_home_after_time but comes home afterswards. This would ensure that after 15 minutes the alarm system gets armed.

**Comment by andyinno** [comment]
Yes, I understand that.

So you do not intend to arm_home if someone is already at home at the time of the trigger?

I mean, the two scenario are different.

NOBODY_HOME => Time trigger => someone gets home => 15mins => alarm_home
SOMEONE_HOME => Time trigger => 15mins => alarm_home

I like the second one to be active. If I am home I prefer that the module triggers alarm_home at some specified time trigger.

**Comment by runningman84** [comment]
Right now if only somone and not everybody is at home the system is not armed.

My assumption was that you do not want to raise an alarm if the second person comes home... because this would most likeley be a false alert.

But I can see that somebody also wants to have it armed in this case too. In my usecase I normally do not use any button or manual control of the alarm system. It just works like it is not even there...

**Comment by andyinno** [comment]
Yes, I understand your point.

Thank you for explaining it.
I think you are right with the scenario you are describing. 
What do you think about this scenario?

Someone at home => time trigger => arm_home
device_tracker sends a trigger not_home => home
system disarms
person enter and arm_home again.

>  In my usecase I normally do not use any button or manual control of the alarm system. It just works like it is not even there...

Yes, me too. I have a button for overriding the defaults if I need it. Like I am home and I want to disarm if someone enters.

    self._device_trackers = self.args.get("device_trackers", [])

I think that if this could accept an array of device_trackers associated with a person it could work with the other case scenario.

like:
    - device_trackers:
        - (indent)person_A:
             - 1) device_tracker.wifi_a
             - 2) device_tracker.bluetooth_a
             - 3) device_tracker.gps_a
         - (indent)person_B:
             - 1) device_tracker.wifi_b
             - 2) device_tracker.bluetooth_b
             - 3) device_tracker.gps_b
Sorry I am unable to format this correctly. I hope it is clear...

At this point you can get if someone registered is entering the house or not.
Or maybe it would be nice to separate the device_tracking in another app and use it for managing home/away?
someway I don't like the way appdaemon and/or home assistant defines home and away states. 
What do you think?

**Comment by runningman84** [comment]
> What do you think about this scenario?
> 
> Someone at home => time trigger => arm_home
> device_tracker sends a trigger not_home => home
> system disarms
> person enter and arm_home again.
> 

I have made this scenario possible. It was just a small change in the code (auto arming if at least one is at home)... This should be a secure and reliable setup because if this setup creates false alerts the normal arm away and return secnario would also fail.

> ```
> self._device_trackers = self.args.get("device_trackers", [])
> ```
> 
> I think that if this could accept an array of device_trackers associated with a person it could work with the other case scenario.
> 
> like:
> - device_trackers:
> - (indent)person_A:
> - 1) device_tracker.wifi_a
> - 2) device_tracker.bluetooth_a
> - 3) device_tracker.gps_a
> - (indent)person_B:
> - 1) device_tracker.wifi_b
> - 2) device_tracker.bluetooth_b
> - 3) device_tracker.gps_b

I suppose you could have this setup right now if you use the person entity instead of a real device tracker entity:
https://www.home-assistant.io/integrations/person/
Please put all your device trackers for a specifc person into a person configuration and use the person entities in the device_trackers list...

**Comment by runningman84** [comment]
> What do you think about this scenario?
> 
> Someone at home => time trigger => arm_home
> device_tracker sends a trigger not_home => home
> system disarms
> person enter and arm_home again.
> 

I have made this scenario possible. It was just a small change in the code (auto arming if at least one is at home)... This should be a secure and reliable setup because if this setup creates false alerts the normal arm away and return secnario would also fail.

> ```
> self._device_trackers = self.args.get("device_trackers", [])
> ```
> 
> I think that if this could accept an array of device_trackers associated with a person it could work with the other case scenario.
> 
> like:
> - device_trackers:
> - (indent)person_A:
> - 1) device_tracker.wifi_a
> - 2) device_tracker.bluetooth_a
> - 3) device_tracker.gps_a
> - (indent)person_B:
> - 1) device_tracker.wifi_b
> - 2) device_tracker.bluetooth_b
> - 3) device_tracker.gps_b

I suppose you could have this setup right now if you use the person entity instead of a real device tracker entity:
https://www.home-assistant.io/integrations/person/
Please put all your device trackers for a specifc person into a person configuration and use the person entities in the device_trackers list...

**Comment by andyinno** [comment]
Hi!
I added the new version yesterday evening and it looks working correctly.
I think that the issue could be closed now. 

Thank you!
